### PR TITLE
Fix pruning logic, cleanup enablement

### LINKED ISSUE #20397: istioctl does not prune gateways anymore
Install the default deployment, then run:

```
ik manifest generate -f <(cat <<EOF 
apiVersion: operator.istio.io/v1alpha1
kind: IstioOperator
spec:
    ingressGateways:
    - enabled: false
EOF
)
```

Output:
```
- Applying manifest for component Base...
✔ Finished applying manifest for component Base.
- Applying manifest for component SidecarInjector...
- Applying manifest for component Pilot...
- Applying manifest for component Addon...
✔ Finished applying manifest for component SidecarInjector.
✔ Finished applying manifest for component Addon.
✔ Finished applying manifest for component Pilot.


✔ Installation complete
```

Note the lack of logs about gateway.

The ingressgateway deployment will remain. If you do the command from a fresh cluster though, no ingress is created, so its just pruning issues not enablement

### PR DESCRIPTION
**Pull request by ostromart**
This fixes https://github.com/istio/istio/issues/20397 but also cleans up and simplifies some enablement logic.

### CODE REVIEW

**Review by howardjohn — APPROVED** [review]

File: operator/pkg/component/component/component.go [hunk 1]
```diff
@@ -82,13 +82,18 @@ type CommonComponentFields struct {
 	// resourceName is the name of all resources for this component.
 	resourceName string
 	// index is the index of the component (only used for components with multiple instances like gateways).
-	index    int
+	index int
+	// spec for the actual gateway, if the component is a gateway.
+	gatewaySpec *v1alpha1.GatewaySpec
```
Comment (richardwxn) [inline operator/pkg/component/component/component.go]: This part doesn't seem to be common for all components? But I understand that this helps to reduce hardcode logic for gateways and components. But just wondering is there better way if possible

File: operator/pkg/component/component/component.go [hunk 2]
```diff
@@ -728,7 +705,43 @@ func createHelmRenderer(c *CommonComponentFields) (helm.TemplateRenderer, error)
...
+	enabled := false
+	switch {
+	case c.componentName.IsCoreComponent():
+		enabled, err = c.Translator.IsComponentEnabled(c.componentName, c.InstallSpec)
+		if err != nil {
+			return false, err
+		}
+
+	case c.componentName.IsGateway():
+		enabled = boolValue(c.gatewaySpec.Enabled)
```
Comment (dgn) [inline operator/pkg/component/component/component.go]: I don't understand this. Why only read from the spec? Why not handle the same way as other components and look at both spec and values?

Comment (ostromart) [reply operator/pkg/component/component/component.go]: good question! we had to push the logic for enablement up to the code where user overlay is applied. the reason is a problem of precedence - we only want the values enablement to override IOS if it came from the user. to simplify things, we translate that value into IOP at that time and use IOP fields only from that point. 
it's really not a great solution so if you can figure out something better here i'd love to improve this.

File: operator/pkg/component/component/component.go (hunk 1 above)
Comment (ostromart) [reply operator/pkg/component/component/component.go]: changed this to interface which i think is a little better. the common go solution in this case is to create an interface and implement it in each struct and store the interface type here but i think this really would be an overkill for this simple situation.

**Pull request merged**